File: Java/src/com/company/DynamicProgramming/LC152_maxProduct.java
package com.company.DynamicProgramming;

public class LC152_maxProduct {
    public int maxProduct(int[] nums) {
        int max_to_cur = nums[0];
        int min_to_cur = nums[0];
        int max = nums[0];
        for(int i = 1; i < nums.length; i++){
            int temp_max = max_to_cur;
            int temp_min = min_to_cur;
            max_to_cur = Math.max(max_to_cur*nums[i], Math.max(nums[i], min_to_cur*nums[i]));
            min_to_cur = Math.min(min_to_cur*nums[i], Math.min(nums[i], temp_max*nums[i]));
            if(max_to_cur > max) max = max_to_cur;
        }
        return max;
    }
}

File: python/DP/LC152_maxProduct.py
    def maxProduct(self, nums: List[int]) -> int:
#         high = nums[0]
        
#         for i in range(len(nums)):
#             res = nums[i]
#             for j in range(i+1, len(nums)):
#                 res *= nums[j]
#                 high = max(res, high)
#             high = max(nums[i], high)
#         return high

        
        dp = [()] * len(nums)
        dp[0] = (nums[0], nums[0])
        val = nums[0]
        
        for i in range(1, len(nums)):
            cur_max = max(nums[i], max(nums[i]*dp[i-1][0], nums[i]*dp[i-1][1]))
            cur_min = min(nums[i], min(nums[i]*dp[i-1][0], nums[i]*dp[i-1][1]))
            dp[i] = (cur_max, cur_min)
            val = max(cur_max, val)
            
        return val
